src/backend/distributed/planner/shard_pruning.c

```diff
@@ -601,6 +575,110 @@ PrunableExpressionsWalker(Node *node, ClauseWalkerContext *context)
 
 
 /*
+ * AddSAOPartitionKeyRestrictionToInstance adds partcol = arrayelem operator
+ * restriction to the current pruning instance for each element of the array. These
+ * restrictions are added to pruning instance to prune shards based on IN/=ANY
+ * constraints.
+ */
+static void
+AddSAOPartitionKeyRestrictionToInstance(ClauseWalkerContext *context,
+										ScalarArrayOpExpr *arrayOperatorExpression)
+{
+	PruningInstance *prune = context->currentPruningInstance;
+	Node *leftOpExpression = linitial(arrayOperatorExpression->args);
+	Node *strippedLeftOpExpression = strip_implicit_coercions(leftOpExpression);
+	bool usingEqualityOperator = OperatorImplementsEquality(
+		arrayOperatorExpression->opno);
+	Expr *arrayArgument = (Expr *) lsecond(arrayOperatorExpression->args);
+
+	/* found partcol = ANY(const, value, s); or partcol IN (const,b,c); */
+	if (usingEqualityOperator && strippedLeftOpExpression != NULL &&
+		equal(strippedLeftOpExpression, context->partitionColumn) &&
+		IsA(arrayArgument, Const))
+	{
+		ArrayType *array = NULL;
+		int16 typlen = 0;
+		bool typbyval = false;
+		char typalign = '\0';
+		Oid element_type = 0;
+		ArrayIterator arrayIterator = NULL;
+		Datum arrayElement = 0;
+		Datum inArray = ((Const *) arrayArgument)->constvalue;
```

Comment:
FWIW, I find that name confusing - we're not even necessarily dealing with an IN() here.